Assert not same executor when completing future

henningandersen opened the pull request:
A common deadlock pattern is waiting and completing a future on the same executor. This only works until the executor is fully depleted of threads. Now assert that waiting for a future to be completed and the completion happens on different executors.

Introduced UnsafePlainActionFuture, used in all offending places, allowing those to be tackled independently.

henningandersen committed:
Assert not same executor when completing future

A common deadlock pattern is waiting and completing a future on the same executor.
This only works until the executor is fully depleted of threads. Now assert that
waiting for a future to be completed and the completion happens on different
executors.

henningandersen commented:
Opening as draft for now to see how CI behaves.

henningandersen committed:
Get proper debug info out of assertion.

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java:
```diff
@@ -266,6 +267,27 @@ public static String threadName(final String nodeName, final String namePrefix)
 
+    // to be used in assertions only.
+    public static boolean differentExecutors(Thread thread1, Thread thread2) {
+        assert thread1 != thread2 : "only call for different threads";
```
Probably want to return `false` here rather than failing with an unhelpful message?

henningandersen replied:
This was sort of on purpose in that, the method should only be used in situations where you are clearly comparing different threads. It seems that if you use this for the same thread your assertion is not written correctly. Hence more of an assertion on the assertion. Not too tied to this though.

DaveCTurner replied:
Oh I see, yeah, we can't be completing on the waiting thread because that thread is waiting...

(deserves a comment to avoid redoing that thinking tho)

DaveCTurner commented on server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java:
```diff
@@ -266,6 +267,27 @@ public static String threadName(final String nodeName, final String namePrefix)
+
+    // visible for tests
+    static String executorName(Thread thread) {
+        assert Assertions.ENABLED;
```
```suggestion
        assert Assertions.ENABLED;
```

This is a no-op: if assertions are disabled, we don't even run this check :)

henningandersen replied:
oh, I fixed that in one place already, thanks.

DaveCTurner reviewed (commented):
Nice idea. I think we can make this even stronger tho, we can still deadlock if we have a loop of executors all waiting on each other to complete. Can we declare up-front all pairs of executors that can be in a block/complete relationship and verify that this permits no such cycles?

henningandersen committed:
Remove dummy assertion.

kingherc commented on server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java:
```diff
@@ -266,6 +267,26 @@ public static String threadName(final String nodeName, final String namePrefix)
+    // visible for tests
+    static String executorName(Thread thread) {
+        String name = thread.getName();
+        int executorNameEnd = name.lastIndexOf(']', name.length() - 2);
```
nit A comment on why `name.length() - 2` would be helpful. Why is it like that? I understand that probably there's another `]` you'd like to skip? But I don't remember how the thread name is formatted out of the top of my head. Reading

```
    public static String threadName(final String nodeName, final String namePrefix) {
        // TODO missing node names should only be allowed in tests
        return "elasticsearch" + (nodeName.isEmpty() ? "" : "[") + nodeName + (nodeName.isEmpty() ? "" : "]") + "[" + namePrefix + "]";
    }
```

and assuming namePrefix is the executor name, I'm not sure where the other `]` would be.

henningandersen replied:
The thread naming is something like `XXXX[executor-name][T#NNN]`. We need to skip the running number part. The index passed into `lastIndexOf` is inclusive.

I added a short comment on this.

fcofdez replied:
Maybe we could use a regex here? maybe that's clearer? (famous last words)

DaveCTurner replied:
This time there's even _two_ obligatory XCKDs:
- https://xkcd.com/208/
- https://xkcd.com/1171/

henningandersen replied:
I prefer the manual parsing done here, seems simplest.

kingherc reviewed (commented):
Nice idea!

henningandersen committed:
cs

henningandersen committed:
comments.

henningandersen commented:
> Nice idea. I think we can make this even stronger tho, we can still deadlock if we have a loop of executors all waiting on each other to complete. Can we declare up-front all pairs of executors that can be in a block/complete relationship and verify that this permits no such cycles?

Sounds like a good idea. For now, I'd like to ensure this can be made to pass CI at all, I think there could be many cases of just this first assertion failing. I might defer your suggestion to a follow-up.

henningandersen commented on server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java:
```diff
@@ -266,6 +267,26 @@ public static String threadName(final String nodeName, final String namePrefix)
     }
 
+    // to be used in assertions only.
+    public static boolean differentExecutors(Thread thread1, Thread thread2) {
```
I had that first, but since it no longer asserts that the executors are different but rather returns whether they are, I renamed it to avoid confusion.

fcofdez reviewed (commented):
This looks good

henningandersen committed:
nasty hack to see more issues.

henningandersen commented on server/src/main/java/org/elasticsearch/index/engine/Engine.java:
```diff
@@ -1964,8 +1964,13 @@ protected boolean blockingAllowed() {
                     || super.blockingAllowed();
             }
         };
+        CountDownLatch latch = new CountDownLatch(1);
         drainOnCloseListener.addListener(future);
+        drainOnCloseListener.addListener(ActionListener.releasing(latch::countDown));
         try {
+            // todo: hack to circumvent same executor check here to see if more failures appear.
+            // need to make this method async, which includes making engine close async, which we should be able to do now.
+            latch.await();
```
These changes will not go in as is, but wanted to get more build failures out.

henningandersen committed:
Add and use UnsafePlainActionFuture

Too many issues to fix in one PR, add a class that is used where we rely on
notifying on same thread to at least have visibility.

henningandersen committed:
CCR unsafety

henningandersen committed:
Completion stats and searchable snapshot

henningandersen committed:
Disruption IT

henningandersen committed:
Transport test cases

henningandersen committed:
Transport test case plus corrupted blob store IT

henningandersen committed:
spotless

henningandersen committed:
right executor

henningandersen committed:
ml

henningandersen committed:
security

henningandersen committed:
Merge remote-tracking branch 'origin/main' into assert_not_same_executor_when_completing_future

henningandersen committed:
Simplifiy

henningandersen committed:
ccr double encounter

henningandersen committed:
silly mistake

henningandersen committed:
inference runner

henningandersen commented on server/src/main/java/org/elasticsearch/action/support/UnsafePlainActionFuture.java:
```diff
@@ -0,0 +1,51 @@
+ * This future is unsafe, since it allows notifying the future on the same thread pool executor that it is being waited on. This
+ * is a common deadlock scenario, since all threads may be waiting and thus no thread may be able to complete the future.
+ */
+public class UnsafePlainActionFuture<T> extends PlainActionFuture<T> {
```
I opted to add this unsafe variant that is then used in all places where I found conflicts, since fixing them all would make this PR size unmanageable. Using this we can now fix them one by one, spreading out the load too.

DaveCTurner replied:
:+1: can we mark this as `@Deprecated(forRemoval = true)` so that IDEs highlight its usages?

kingherc replied:
Do we need a task/issue/ticket on fixing them one-by-one?

henningandersen replied:
I think one by one is too granular. I'll probably create area level ones (perhaps a bit more granular, for instance the one in `AbstractClient` needs a separate one).

henningandersen commented:
This now has a clean build so marking ready for review. I'll do more CI work to scrape up more occurrences before merge anyway (which can result in more unsafe future usages).

DaveCTurner commented on server/src/main/java/org/elasticsearch/client/internal/support/AbstractClient.java:
```diff
@@ -410,7 +411,13 @@ protected <Request extends ActionRequest, Response extends ActionResponse> void
-    private static class RefCountedFuture<R extends RefCounted> extends PlainActionFuture<R> {
+    // todo: the use of UnsafePlainActionFuture here is quite broad, we should find a better way to be more specific
+    // (unless making all usages safe is easy).
+    private static class RefCountedFuture<R extends RefCounted> extends UnsafePlainActionFuture<R> {
```
Yikes. I suspect we can/should move all usages of this into tests, and do the ref-counting (and asyncification) properly in prod code. But I see that's not a small change. Ok for now...

henningandersen replied:
Yeah, I also pondered on this for a short while, but decided to defer this issue for now. I think I agree to move to all prod client interactions being async now. If it is too big we would need to stop notifying on generic thread pool (one more pool maybe, hopefully as an interim step). We can discuss more when we tackle it.

DaveCTurner reviewed (approved):
Great stuff, I like it.

kingherc reviewed (approved):
LGTM as long as CI is happy

tlrx reviewed (approved):
LGTM, nice idea!

henningandersen committed:
Mark deprecated.

henningandersen committed:
Stateless prewarming

henningandersen committed:
spotless

henningandersen committed:
Disable assertion for merge.

henningandersen committed:
Merge remote-tracking branch 'origin/main' into assert_not_same_executor_when_completing_future
